backend/pkg/serde/service.go

```diff
@@ -0,0 +1,282 @@
+		})
+	}
+
+	addTS := opts.Troubleshoot
+	if rp == nil || err != nil || (rp != nil && rp.Encoding == "") {
+		// Anything else is considered binary
+		rp = &RecordPayload{
+			Encoding: PayloadEncodingBinary,
+		}
+
+		// if we failed to find a successful serde always add troubleshooting info
+		addTS = true
+	}
+
+	rp.PayloadSizeBytes = len(payload)
+	rp.IsPayloadNull = payload == nil
+
+	if opts.IncludeRawData {
+		rp.OriginalPayload = payload
+	}
+
+	if len(payload) > opts.MaxPayloadSize {
+		rp.IsPayloadTooLarge = true
+		rp.NormalizedPayload = nil
+	}
+
+	if addTS {
+		rp.Troubleshooting = troubleshooting
+	}
+
+	return rp
+}
+
+// DeserializationOptions that can be provided by the requester to influence
+// the deserialization.
+type DeserializationOptions struct {
+	// KeyEncoding may be specified by the frontend to indicate that this
+	// encoding type shall be used to deserialize the key. This is helpful,
+	// if the requester knows that a primitive type like int16 is used, which couldn't
+	// be guessed automatically. No other encoding method will be tried if this
+	// method has failed. Troubleshoot should always be set to true in this case.
+	KeyEncoding PayloadEncoding
+
+	// PreferredValueEncoding may be specified by the frontend to indicate that this
+	// encoding type shall be used to deserialize the value. This is helpful,
+	// if the requester knows that a primitive type like int16 is used, which couldn't
+	// be guessed automatically. No other encoding method will be tried if this
+	// method has failed. Troubleshoot should always be set to true in this case.
+	ValueEncoding PayloadEncoding
+
+	// MaxPayloadSize is the maximum size of the payload.
+	MaxPayloadSize int
+
+	// Troubleshoot can be enabled to return additional information which reports
+	// why each performed deserialization strategy has failed. If the first
+	// tested encoding method worked successfully no troubleshooting information
+	// is returned.
+	Troubleshoot bool
+
+	// IncludeRawData can be enabled to include raw binary data in the returned output.
+	IncludeRawData bool
+}
+
+// SerializeRecord will serialize the input.
+func (s *Service) SerializeRecord(ctx context.Context, input SerializeInput) (*SerializeOutput, error) {
+	keySerResult := RecordPayloadSerializeResult{}
+	valueSerResult := RecordPayloadSerializeResult{}
+
+	var err error
+	sr := SerializeOutput{}
+	sr.Key = &keySerResult
+	sr.Value = &valueSerResult
+
+	// key
+	if input.Topic != "" {
+		input.Key.Options = append(input.Key.Options, WithTopic(input.Topic))
+	}
+
+	keyTS := make([]TroubleshootingReport, 0)
+	found := false
+	for _, serde := range s.SerDes {
+		if input.Key.Encoding != serde.Name() {
+			continue
+		}
+
+		found = true
+
+		bytes, serErr := serde.SerializeObject(ctx, input.Key.Payload, PayloadTypeKey, input.Key.Options...)
+		if serErr == nil {
+			err = nil
+			keySerResult.Encoding = serde.Name()
+			keySerResult.Payload = bytes
+			break
+		}
+		err = serErr
```

Comment:
```suggestion
		var bytes []byte // I assume?
		bytes, err = serde.SerializeObject(ctx, input.Key.Payload, PayloadTypeKey, input.Key.Options...)
		if err == nil {
			keySerResult.Encoding = serde.Name()
			keySerResult.Payload = bytes
			break
		}
```

Reply:
Maybe also worth to refactor it so that you just find the serde in that for loop and proceed outside the for loop with the serialization of the found serde